src/server/frontend/wayland/wayland_connector.cpp

```diff
@@ -2186,6 +2186,50 @@ class WlShell : public wayland::Shell
     std::shared_ptr<mf::Shell> const shell;
     WlSeat& seat;
 };
+
+struct DataDevice : wayland::DataDevice
+{
+    DataDevice(struct wl_client* client, struct wl_resource* parent, uint32_t id) :
+        wayland::DataDevice(client, parent, id)
+    {
+    }
+
+    void start_drag(
+        std::experimental::optional<struct wl_resource*> const& source, struct wl_resource* origin,
+        std::experimental::optional<struct wl_resource*> const& icon, uint32_t serial) override
+    {
+        (void)source, (void)origin, (void)icon, (void)serial;
+    }
+
+    void set_selection(std::experimental::optional<struct wl_resource*> const& source, uint32_t serial) override
+    {
+        (void)source, (void)serial;
+    }
+
+    void release() override
+    {
+    }
+};
+
+struct DataDeviceManager : wayland::DataDeviceManager
+{
+    DataDeviceManager(struct wl_display* display, uint32_t max_version) :
```

Comment:
I don't think that max_version should be an externally-visible constructor argument; the code in DataDeviceManager either supports a given protocol version, or it doesn't, and we always have to support clients binding to a lower version than we support, too.

I think we should strive for the protocol-implementing objects to contain the versioning differences as much as possible.